### PR TITLE
Select timer count direction (#1926)

### LINKED ISSUE #1926: Timer count direction option?
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

Currently if we do not define a "start" time in the timers they count upward.(no notification when time is up because no time is set in "start")
If we define a "start" time in the timers they count downward and prevent me when the time set in “start” is past.

Personally, when I fly, I don't find this count down intuitive (you have to do a subtraction to find out how much time has passed) 
But I like to be notified when the time is up.

### Describe the solution you'd like

Would it be possible to have an option that would allow us to define in which direction the timer counts? (Count up / count down)

This would allow us to have a more intuitive timer (=> Count up)  AND be notified when the time is up.

### Describe alternatives you've considered

Currently I'm using logical switches to get the job done, but it's also not intuitive to set the "timer time" in the logical switch.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by eshifri**
This PR allows user to select timer direction: display/announce elapsed time or remaining time.

Fixes #1926

Summary of changes:  add UI for BW and color radio to select direction of the timer counter.
Support TX and Companion. 
Based on the main after lvgl merge.

### CODE REVIEW

File: companion/src/firmwares/edgetx/yaml_modeldata.cpp
Comment (raphaelcoeffic) [inline companion/src/firmwares/edgetx/yaml_modeldata.cpp:301]: Is the extra `T` on purpose?

Comment (eshifri) [reply companion/src/firmwares/edgetx/yaml_modeldata.cpp:301]: Nope. Must be slip of a finger.

File: radio/src/gui/colorlcd/widgets/timer.cpp
Comment (raphaelcoeffic) [inline radio/src/gui/colorlcd/widgets/timer.cpp:76]: Debug code?

Comment (eshifri) [reply radio/src/gui/colorlcd/widgets/timer.cpp:76]: Yes! Thank you for catching!

File: radio/src/storage/yaml/yaml_datastructs_nv14.cpp
Comment (raphaelcoeffic) [inline radio/src/storage/yaml/yaml_datastructs_nv14.cpp:469]: Please do not use formatting tools on generated code. This will generate changes for the next one re-generating this code.

File: radio/src/translations/cn.h
Comment (raphaelcoeffic) [inline radio/src/translations/cn.h:231]: “\0” are not necessary anymore.

File: radio/src/translations/es.h
Comment (raphaelcoeffic) [inline radio/src/translations/es.h:33]: This looks suspicious. This file should have as few changes as the others.

File: radio/src/translations/cn.h
Comment (eshifri) [reply radio/src/translations/cn.h:231]: OK.

File: radio/src/translations/fr.h
Comment (raphaelcoeffic) [inline radio/src/translations/fr.h:274]: No LEN_… anymore please, they are not used anymore.

**Review by raphaelcoeffic — COMMENTED** [review]
Please re-check the translation files.

**Comment by pfeerick** [comment]
Testing:
- [X] X9D+2018 - elapsed, remaining, companion sync (both ways)
- [ ] TX12 - elapsed, remaining, companion sync (both ways)
- [ ] TX16S - elapsed, remaining, companion sync (both ways)

**Review by pfeerick — APPROVED** [review]
I'm happy with this now. 

@elecpower Can you give this a quick look over and see if you're happy with the Companion side changes... everything seems to be working and nothing stands out to me. 

image

**Comment by pfeerick** [comment]
@elecpower I'd like to merge this in tomorrow if there's no issues. Sorry to bug, it's just that I can't really start on the next PR until this is in as they each need rebasing/updating as they all touch the radio storage yaml.

**Comment by elecpower** [comment]
> @elecpower I'd like to merge this in tomorrow if there's no issues. Sorry to bug, it's just that I can't really start on the next PR until this is in as they each need rebasing/updating as they all touch the radio storage yaml. 

I'll have a gander today

**Comment by elecpower** [comment]
@pfeerick @eshifri  Reviewed and there is a bug with triggers for when to enable/disable the combo box. I'll write a fix and test and pass to @pfeerick to have the final approval so we can get it in tomorrow.

**Comment by elecpower** [comment]
@eshifri @pfeerick I have pushed a commit to fix the enable/disable triggers. Whilst testing I noticed a UI inconsistency between Companion and radio. When time value is zero Companion disables the Show Elapsed combo box however the radio does not. So one of them needs fixing. I leave that to @eshifri.
@eshifri for future reference once a data field is linked to an auto widget the widget takes over updating the data field so no need for on field changed. You connect to the auto widget currentDataChanged if you need to trigger other stuff.

**Comment by MervDale** [comment]
I meant to ask this question ages ago but is this a problem on some radios because I have never had this issue ....If you set a flight time, say 6 minutes for example the the radio timer will count down to zero from 6 minutes but if you do not set a time then it will count up to infinity but call out the minutes as it it times up.?

**Comment by eshifri** [comment]
Please take a look at the feature request. I think what the requester wanted is the count to infinity with the "timer expired" warning and the change of the color. Personally, I do not find it very useful, but why not if people want it?

**Comment by pfeerick** [comment]
The problem therein is it counts to infinity, rather than having a fixed end point (and then rollover). The person may be doing say competition flying, and wants to know the elapsed time, _not_ the remaining time (as having to do the inversion in your head unnecessarily adds to the workload). i.e. is it F3K that has positive (time elapsed) intervals, and also a fixed end time period?

**Comment by MervDale** [comment]
> The problem therein is it counts to infinity, rather than having a fixed end point (and then rollover). The person may be doing say competition flying, and wants to know the elapsed time, _not_ the remaining time (as having to do the inversion in your head unnecessarily adds to the workload). i.e. is it F3K that has positive (time elapsed) intervals, and also a fixed end time period?

So here is a question for you, when we add a  positive number to say timer 2 it will count DOWN, when we put no number it will count UP but is was madei t  possible to put a negative number in the timer would it count UP the same way as it does when counting DOWN?

**Comment by pfeerick** [comment]
If I understand your question, then essentially yes. However, that would require some a few changes to the timer system and testing to ensure nothing else broke, as the timers have been designed on the basis of no negative numbers (i.e. uint values). It is much simpler (and more intuitive IMO) to have an option (like this introduces) that establishes whether are you counting from zero to a set value, or from a set value to zero.

**Comment by pfeerick** [comment]
> Whilst testing I noticed a UI inconsistency between Companion and radio. When time value is zero Companion disables the Show Elapsed combo box however the radio does not. So one of them needs fixing.

@elecpower Was this *just* for the TX16S / colorlcd radios? As when I was testing it on X9D+2019 and TX12S it did hide the settings when timer was set to zero. I'll see if we can do the same on the colorlcd radios, but it may be a necessary quirk atm.

**Comment by elecpower** [comment]
> > Whilst testing I noticed a UI inconsistency between Companion and radio. When time value is zero Companion disables the Show Elapsed combo box however the radio does not. So one of them needs fixing.
> 
> @elecpower Was this _just_ for the TX16S / colorlcd radios? As when I was testing it on X9D+2019 and TX12S it did hide the settings/option when timer was set to zero. I'll see if we can do the same on the colorlcd radios, but it may be a necessary quirk atm.

@pfeerick observed on TX16S

**Comment by pfeerick** [comment]
Perfect... e3bf23e should resolve that then. If you and @eshifri are happy with the current state this is right for the next nightly...

**Comment by MervDale** [comment]
This relates to the nightlies I have noticed when running a timer widget if you go back into the timer page," just in and out no mods it will reset the timer back to zero  does not do this in 2.7.1( different setup) I found this trying to help out a user that was having issues with Timer 3 zeroing when he powered down in 2.7.1  most likely did not have persistent or manual reset option(s) active

**Comment by elecpower** [comment]
Looks okay

**Comment by pfeerick** [comment]
> This relates to the nightlies I have noticed when running a timer widget if you go back into the timer page," just in and out no mods it will reset the timer back to zero does not do this in 2.7.1( different setup) I found this trying to help out a user that was having issues with Timer 3 zeroing when he powered down in 2.7.1 most likely did not have persistent or manual reset option(s) active

@MervDale Ok, that is completely different, and looks like a regression... timers were fixed in 2.6 or 2.7 so that an *edit* would reset the timer (as they weren't before, and you'd get some interesting "timer 1 elapsed" or whatever and would have to reset it to make it see reason), but as you point out here, merely visiting the timer page now is enough to reset the timer start value... which shouldn't be the case. I'm guessing  `timerValueUpdater` is being fired when the current value is obtained, rather than on a change. Can you make an issue for that... basically `main` is current setting the timer start value whenever the page is entered rather than on edit. The callback I just added may be of some help, but this is out of scope for this PR.

**Comment by eshifri** [comment]
I have no problems with this of course.
Strange: I'm sure I have tested this use case in my implementation and it worked. Whatever...

**Comment by pfeerick** [comment]
@eshifri Never assume I write good code.. especially with LVGL stuff ;) It'll end badly for both of us! 😆 The B&W radios were fine, the colorlcd UI didn't make any attempts to hide the option. Anyway, it's up to the nightly users to gripe about it now. Looks like there's a few more tweaks needed to the new timer UI anyway (i.e. Merv's timer reset bug mentioned above, and there's some translations strings to be added.

**Comment by elecpower** [comment]
I also came across @MervDale bug or maybe a slight variant in testing this PR. When starting the sim I received a time elapsed verbal on a number of occasions and then it went away. So I'm guessing the check occurs before the starting value is set and it is checking against whatever happens to be in the memory location on start.

**Comment by pfeerick** [comment]
That sounds like another issue (around timers still), not sure what other changes were made around there so will add to the list ;)

**Pull request merged**
